### PR TITLE
Add "full_outer" name to join types

### PR DESCRIPTION
**Pull request by BartekH**
I have discovered that "full_outer" name option is working in Spark 2.0, but it is not printed in exception. Please verify.

## What changes were proposed in this pull request?

(Please fill in changes proposed in this fix)

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)
(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

**Comment by wzhfy** [comment]
"full_outer" will be replaced by "fullouter" after `.replace("_", "")`, right?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/joinTypes.scala [hunk 1]
```diff
@@ -24,7 +24,7 @@ import org.apache.spark.sql.catalyst.expressions.Attribute
 object JoinType {
   def apply(typ: String): JoinType = typ.toLowerCase(Locale.ROOT).replace("_", "") match {
     case "inner" => Inner
-    case "outer" | "full" | "fullouter" => FullOuter
+    case "outer" | "full" | "fullouter" | "full_outer" => FullOuter
```
Comment (sameeragarwal) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/joinTypes.scala]: as @wzhfy points out, given that we remove underscores, this matching case is unnecessary. The same is true for `left_outer`, `right_outer` and `left_semi`.

**Comment by BartekH** [comment]
That's ok, but "full_outer" is invisible in exception information. Please consider add it, because is is possible to use "full_outer". When wrong name is provided and exception occurs, there is no information about "full_outer" option. It's a little bit misleading.

**Comment by wzhfy** [comment]
@BartekH Yes, I think we can add that to exception message. Please also add a test case for checking supported join types.

**Comment by HyukjinKwon** [comment]
Ping @BartekH, how is it going?

**Comment by BartekH** [comment]
I'll make pull request today. Sorry for delay

**Comment by HyukjinKwon** [comment]
Could we just fold the changes in 18219 to here?

**Comment by BartekH** [comment]
@ HyukjinKwon Done.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/JoinTypesTest.scala [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+
+package org.apache.spark.sql.catalyst.plans
+
+import org.apache.spark.SparkFunSuite
+
+class JoinTypesTest extends SparkFunSuite {
+
+  test("construct an Inner type") {
+    assert(JoinType.apply("inner") === Inner)
```
Comment (sameeragarwal) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/JoinTypesTest.scala]: nit: can just have `JoinType("...")` here and below. Also, can we have a test that checks for the correct exception for wrong join types? Something along the lines of:

```scala
val ex = intercept[IllegalArgumentException] {
  JoinType.apply("wrong_join")
}
ex.getMessage() should contain "..."
```

**Comment by sameeragarwal** [comment]
ok to test

**Comment by BartekH** [comment]
Please rerun jenkins job.

**Comment by gatorsmile** [comment]
retest this please

**Comment by BartekH** [comment]
It has failed again. It's not my fault, please retest it one more time.

**Comment by gatorsmile** [comment]
retest this please

**Comment by wzhfy** [comment]
retest this please

**Comment by BartekH** [comment]
Does your Jenkins work properly? It failed again, and again cause is different.

**Comment by gatorsmile** [comment]
test this please

**Comment by gatorsmile** [comment]
retest this please

**Comment by gatorsmile** [comment]
retest this please

**Comment by gatorsmile** [comment]
LGTM

**Comment by gatorsmile** [comment]
Thanks! Merging to master.

**Pull request closed without merging**